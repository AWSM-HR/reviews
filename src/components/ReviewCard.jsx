import React from 'react';
import Proptypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardMedia from '@material-ui/core/CardMedia';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import Avatar from '@material-ui/core/Avatar';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import { red } from '@material-ui/core/colors';
import FiberManualRecordIcon from '@material-ui/icons/FiberManualRecord';
import ThumbUpAltOutlinedIcon from '@material-ui/icons/ThumbUpAltOutlined';
import ShareIcon from '@material-ui/icons/Share';
import MoreHorizIcon from '@material-ui/icons/MoreHoriz';
import Carousel from 'react-material-ui-carousel';
import Rating from '@material-ui/lab/Rating';
import LocationOnIcon from '@material-ui/icons/LocationOn';
import moment from 'moment';

const useStyles = makeStyles((theme) => ({
  root: {
    maxWidth: '60%',
  },
  media: {
    height: 0,
    paddingTop: '56.25%', // 16:9
  },
  expand: {
    transform: 'rotate(0deg)',
    marginLeft: 'auto',
    transition: theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    }),
  },
  expandOpen: {
    transform: 'rotate(180deg)',
  },
  avatar: {
    backgroundColor: red[500],
  },
  iconFilled: {
    color: '#34E0A1',
  },
}));

const ReviewCard = ({ review, helpfulClickHandler }) => {
  const classes = useStyles();
  const date = moment(review.created_at).format('MMM-YYYY');
  const dateOfExperience = moment(review.dateOfExperience).format('MMM-YYYY');
  const { _id } = review;

  const imagesComponent = () => {
    const { images } = review;
    if (images.length === 0) {
      return null;
    }
    if (images.length === 1) {
      return (
        <CardMedia
          className={classes.media}
          image={images[0]}
          title="reviews pic"
        />
      );
    }
    return (
      <Carousel>
        {images.map((image) => (
          <CardMedia
            className={classes.media}
            image={image}
            title="reviews pic"
          />
        ))}
      </Carousel>
    );
  };

  const helpfulVotes = () => {
    if (review.helpfulVotes === 1) {
      return '1 Helpful vote.';
    }
    return `${review.helpfulVotes} Helpfules votes.`;
  };

  return (
    <Card className={classes.root}>
      <CardHeader
        avatar={
          (
            <Avatar aria-label="recipe" className={classes.avatar}>
              <img src={review.profilePic} alt="profile pic" />
            </Avatar>
          )
        }
        action={
          (
            <IconButton aria-label="settings">
              <MoreHorizIcon />
            </IconButton>
          )
        }
        title={(
          <Typography variant="subtitle1" color="textPrimary" component="div">
            {review.userName}
            <Typography display="inline" variant="caption" color="textSecondary" component="p">
              {` wrote a review ${date}`}
            </Typography>
          </Typography>
        )}
        subheader={(
          <>
            <LocationOnIcon style={{ fontSize: 14 }} />
            <Typography display="inline" variant="caption" color="textSecondary" component="p">
              {review.userHomeLocation}
            </Typography>
          </>
        )}
      />
      {imagesComponent()}
      <CardContent>
        <Rating
          max={5}
          value={review.starRating}
          readOnly
          className={classes.iconFilled}
          icon={<FiberManualRecordIcon styles={{ color: '#34E0A1' }} />}
        />
        <Typography variant="h6" color="textPrimary" component="h6">
          {review.reviewTitle}
        </Typography>
        <Typography variant="body2" color="textSecondary" component="p" paragraph>
          {review.reviewBody}
        </Typography>
        <Typography variant="body2" color="textPrimary" component="div" paragraph>
          Date of experience:
          <Typography display="inline" variant="body2" color="textSecondary" component="p">
            {` ${dateOfExperience}`}
          </Typography>
        </Typography>
        <Typography variant="caption" color="textSecondary" component="p" paragraph>
          {helpfulVotes()}
        </Typography>
      </CardContent>
      <CardActions disableSpacing>
        <IconButton aria-label="i found this helpful" onClick={(e) => { helpfulClickHandler(e); }} data-id={_id}>
          <ThumbUpAltOutlinedIcon data-id={_id} />
          <Typography variant="caption" color="textPrimary" component="p" data-id={_id}>
            Helpful
          </Typography>
        </IconButton>
        <IconButton aria-label="share">
          <ShareIcon />
          <Typography variant="caption" color="textPrimary" component="p">
            Share
          </Typography>
        </IconButton>
      </CardActions>
    </Card>
  );
};

ReviewCard.propTypes = {
  review: Proptypes.shape({
    created_at: Proptypes.string,
    dateOfExperience: Proptypes.string,
    destination: Proptypes.string,
    helpfulVotes: Proptypes.number,
    images: Proptypes.arrayOf(Proptypes.string),
    profilePic: Proptypes.string,
    reviewBody: Proptypes.string,
    reviewTitle: Proptypes.string,
    starRating: Proptypes.number,
    userHomeLocation: Proptypes.string,
    userName: Proptypes.string,
    _id: Proptypes.string,
  }).isRequired,
  helpfulClickHandler: Proptypes.func.isRequired,
};

export default ReviewCard;

/*

<CardMedia
  className={classes.media}
  image={image}
  title="reviews pic"
/>

*/
