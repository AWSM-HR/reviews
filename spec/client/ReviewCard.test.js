import React from 'react';
import Enzyme, {mount} from 'enzyme';
import ReviewCard from '../../src/components/ReviewCard';
import { createMount, createShallow, createRender } from '@material-ui/core/test-utils';
import toJSON from 'enzyme-to-json';


const review = {
  created_at: "2021-01-16T01:04:01.548Z",
  dateOfExperience: "2021-01-16T01:04:01.548Z",
  destination: "Bangkok",
  helpfulVotes: 11,
  images: ["http://d20lp9tw1uk7y6.cloudfront.net/images/tripadvisor_thailand_4.jpg", "http://d20lp9tw1uk7y6.cloudfront.net/images/tripadvisor_thailand_82.jpg", "http://d20lp9tw1uk7y6.cloudfront.net/images/tripadvisor_thailand_23.jpg", "http://d20lp9tw1uk7y6.cloudfront.net/images/tripadvisor_thailand_46.jpg"],
  profilePic: "http://d20lp9tw1uk7y6.cloudfront.net/profilePics/profilePic_people_1.jpg",
  reviewBody: "Et facere aliquam sit eveniet. Ullam deserunt eveniet quos distinctio. Velit nostrum voluptatibus sapiente aut totam molestiae inventore dignissimos porro. Voluptatem nihil qui iste doloribus natus esse. Excepturi praesentium quo earum. Laudantium aliquid voluptas totam enim nostrum.",
  reviewTitle: "dolores vitae quis",
  starRating: 1,
  userHomeLocation: "Niger",
  userName: "Herta Corwin",
}

describe('ReviewCard component', () => {
  let mount;
  let shallow;

  beforeEach(() => {
    mount = createMount();
    shallow = createShallow();
  });


  test('Renders the correct helpful votes count.', () => {
    const wrapper = shallow(<ReviewCard review={review} helpfulClickHandler={() => null}/>);

    expect(wrapper.exists()).toBe(true);
  });

  test('Increments helpful votes count on button click', () => {
    const wrapper = mount(<ReviewCard review={review} helpfulClickHandler={() => null}/>);

    expect(wrapper.find('p[data-testid="votesCounter"]').text()).toBe('11 Helpful votes.');
  });

  test('Renders the correct amout of stars for a review rating', () => {
    const wrapper = mount(<ReviewCard review={review} helpfulClickHandler={() => null}/>);
    expect(wrapper.find('span.MuiRating-iconEmpty').length).toBe(4);
    expect(wrapper.find('span.MuiRating-iconFilled').length).toBe(1);
  });

  test('Renders date of review in a user readable format', () => {
    const wrapper = mount(<ReviewCard review={review} helpfulClickHandler={() => null}/>);
    expect(wrapper.find('p[data-testid="dateOfReview"]').text()).toBe(' wrote a review Jan-2021');
  })
});
