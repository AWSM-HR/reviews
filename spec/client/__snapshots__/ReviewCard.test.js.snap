// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewCard component ReviewCard matches snapshot 1`] = `[Function]`;
